gene_expression_tasks.py

```diff
@@ -0,0 +1,622 @@
+                shape=(len(gene_expr_sub), len(batch_joinids)),
+                dtype=np.float32
+            ).toarray()
+        else:
+            batch_matrix = np.zeros((len(gene_expr_sub), len(batch_joinids)), dtype=np.float32)
+
+        with warnings.catch_warnings():
+            warnings.filterwarnings("ignore")
+            for i, gene_id in enumerate(batch_genes):
+                if np.var(batch_matrix[:, i], ddof=1) > 0:
+                    try:
+                        corr, p_value = pearsonr(gene_expr_sub, batch_matrix[:, i])
+                        if p_value < 0.05 and not np.isnan(p_value):
+                            local_correlations[gene_id] = corr
+                    except:
+                        continue
+
+        return local_correlations
+
+def get_coexpression_matrix_for_tissue(gene, tissue_uberon_id, cell_type=None, k=500):
+    
+    def get_coexpression_matrix(gene, tissue_uberon_id, cell_type, k=500, batch_size=1000, use_prefilter=False):
+        with cellxgene_census.open_soma(census_version="2024-07-01") as census:
+            experiment = census["census_data"]["homo_sapiens"]
+            
+            # Use tissue_ontology_term_id to filter by UBERON ID
+            if cell_type:
+                # Filter by both tissue and cell type
+                value_filter = f"tissue_ontology_term_id == '{tissue_uberon_id}' and cell_type == '{cell_type}'"
+            else:
+                # Filter by tissue only
+                value_filter = f"tissue_ontology_term_id == '{tissue_uberon_id}'"
+            
+            try:
+                axis_query = experiment.axis_query(
+                    measurement_name="RNA",
+                    obs_query=soma.AxisQuery(value_filter=value_filter)
+                )
+            except ValueError:
+                logger.warning(f"No cells found for tissue UBERON ID '{tissue_uberon_id}'")
+                return [], [], []
+
+            obs_joinids = axis_query.obs_joinids().to_numpy()
+
+            if len(obs_joinids) > 100000:
+                obs_joinids = obs_joinids[:100000] 
+                n = 100000
+            else:
+                n = len(obs_joinids)
+                
+            if n == 0:
+                logger.warning(f"No cells found for tissue UBERON ID '{tissue_uberon_id}'")
+                return [], [], []
+
+            var_df = experiment.ms["RNA"].var.read(
+                column_names=["soma_joinid", "feature_id"]
+            ).concat().to_pandas().set_index("feature_id")
+            genes = var_df.index.tolist()
+
+            if gene not in genes:
+                logger.warning(f"Gene of interest '{gene}' not found in dataset")
+                return [], [], genes
+
+            gene_joinid = var_df.loc[gene]["soma_joinid"]
+
+            # Get gene expression
+            gene_table_iter = experiment.ms["RNA"].X["raw"].read((obs_joinids.tolist(), [gene_joinid])).tables()
```

Comment:
You should use the `log1p` expression values instead of the `raw` as the raw counts are affected by library size and depth. `log1p` is the normalized version of the raw expression counts with the effect of outliers removed.